File: app/udptransport/src/main/java/org/exthmui/share/udptransport/packets/ResendRequestPacket.java
package org.exthmui.share.udptransport.packets;

import android.util.Log;

import org.apache.commons.lang3.ArrayUtils;
import org.exthmui.share.udptransport.ByteUtils;
import org.exthmui.share.udptransport.Constants;

import java.net.DatagramPacket;
import java.nio.ByteBuffer;
import java.nio.ShortBuffer;
import java.util.Arrays;

public final class ResendRequestPacket extends AbstractCommandPacket<ResendRequestPacket> {
    public static final int MIN_DATA_LENGTH = 1;

    public static final ByteUtils.Tip GROUP_ID_TIP = new ByteUtils.Tip(0, 0);

    public static final ByteUtils.Tip PACKET_IDS_TIP = new ByteUtils.Tip(1);

    public ResendRequestPacket(DatagramPacket packet) {
        super(packet);
    }

    public ResendRequestPacket() {
        super(Constants.UdpCommand.PACKET_RESEND_REQ, MIN_DATA_LENGTH);
    }

    @Override
    public void legalCheck() {
        super.legalCheck();
        if (getCommand() != Constants.UdpCommand.PACKET_RESEND_REQ ||
                super.getData().length < MIN_DATA_LENGTH) {
            Log.e(super.toString(), String.format("Illegal data: %s", Arrays.toString(toDatagramPacket().getData())));
            throw new IllegalArgumentException();
        }
    }

    public ResendRequestPacket setData(byte groupId, byte[] idsBytes) {
        setData(ArrayUtils.addFirst(idsBytes, groupId));
        return this;
    }

    public ResendRequestPacket setData(byte groupId, short[] ids) {
        ByteBuffer byteBuffer = ByteBuffer.allocate(ids.length * 2);

        for (short id : ids) byteBuffer.putShort(id);
        setData(groupId, byteBuffer.array());
        return this;
    }

    public ResendRequestPacket setGroupId(byte groupId) {
        return setData(groupId, getPacketIds());
    }

    public byte getGroupId() {
        return cutDataByTip(GROUP_ID_TIP)[0];
    }

    public ResendRequestPacket setPacketIds(short[] ids) {
        return setData(getGroupId(), ids);
    }

    public short[] getPacketIds() {
        ShortBuffer shortBuffer = ByteBuffer.wrap(getPacketIdsBytes()).asShortBuffer();
        short[] packetIds = new short[shortBuffer.remaining()];
        shortBuffer.get(packetIds);
        return packetIds;
    }

    private byte[] getPacketIdsBytes() {
        return cutDataByTip(PACKET_IDS_TIP);
    }

    @Override
    public String toString() {
        return "ResendRequestPacket{Ids:" + Arrays.toString(getPacketIds()) + "}";
    }
}

File: app/udptransport/src/main/java/org/exthmui/share/udptransport/packets/ResendRequestPacket.kt
package org.exthmui.share.udptransport.packets

import android.util.Log
import org.apache.commons.lang3.ArrayUtils
import org.exthmui.share.udptransport.ByteUtils.Tip
import org.exthmui.share.udptransport.Constants
import java.net.DatagramPacket
import java.nio.ByteBuffer
import java.util.Arrays

class ResendRequestPacket : AckablePacket<ResendRequestPacket?> {
    constructor(packet: DatagramPacket?) : super(packet!!)
    constructor() : super(Constants.UdpCommand.PACKET_RESEND_REQ, MIN_DATA_LENGTH)

    override fun legalCheck() {
        super.legalCheck()
        if (command != Constants.UdpCommand.PACKET_RESEND_REQ ||
            super.data!!.size < MIN_DATA_LENGTH
        ) {
            Log.e(
                super.toString(),
                String.format("Illegal data: %s", Arrays.toString(toDatagramPacket().data))
            )
            throw IllegalArgumentException()
        }
    }

    public override fun clone(): ResendRequestPacket {
        return ResendRequestPacket(toDatagramPacket())
    }

    override fun isAck(): Boolean {
        TODO("Not yet implemented")
    }

    override fun reverseAck(): AckablePacket<ResendRequestPacket?> {
        TODO("Not yet implemented")
    }

    fun setData(groupId: Byte, idsBytes: ByteArray?): ResendRequestPacket {
        setData(ArrayUtils.addFirst(idsBytes, groupId))
        return this
    }

    fun setData(groupId: Byte, ids: ShortArray): ResendRequestPacket {
        val byteBuffer = ByteBuffer.allocate(ids.size * 2)
        for (id in ids) byteBuffer.putShort(id)
        setData(groupId, byteBuffer.array())
        return this
    }

    fun setGroupId(groupId: Byte): ResendRequestPacket {
        return setData(groupId, packetIds)
    }

    val groupId: Byte
        get() = cutDataByTip(GROUP_ID_TIP)[0]

    fun setPacketIds(ids: ShortArray): ResendRequestPacket {
        return setData(groupId, ids)
    }

    val packetIds: ShortArray
        get() {
            val shortBuffer = ByteBuffer.wrap(packetIdsBytes).asShortBuffer()
            val packetIds = ShortArray(shortBuffer.remaining())
            shortBuffer[packetIds]
            return packetIds
        }
    private val packetIdsBytes: ByteArray
        private get() = cutDataByTip(PACKET_IDS_TIP)

    override fun toString(): String {
        return "ResendRequestPacket{Ids:" + Arrays.toString(packetIds) + "}"
    }

    companion object {
        const val MIN_DATA_LENGTH = 1
        val GROUP_ID_TIP = Tip(0, 0)
        val PACKET_IDS_TIP = Tip(1)
    }
}
